xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -472,14 +518,19 @@ private void startRpcStream() {
   }
 
   /**
-   * Handles LDS response to find the HttpConnectionManager message for the requested resource name.
+   * If listenerWatcher is set it just calls handleLdsResponseForListener. Otherwise
+   * handles LDS response to find the HttpConnectionManager message for the requested resource name.
    * Proceed with the resolved RouteConfiguration in HttpConnectionManager message of the requested
    * listener, if exists, to find the VirtualHost configuration for the "xds:" URI
    * (with the port, if any, stripped off). Or sends an RDS request if configured for dynamic
    * resolution. The response is NACKed if contains invalid data for gRPC's usage. Otherwise, an
    * ACK request is sent to management server.
    */
   private void handleLdsResponse(DiscoveryResponse ldsResponse) {
```

Comment:
Let's make this more elegant. Put the logic of unpacking `Listener`s and selecting the requested `Listener` in `handleLdsResponse()`. Then branching into different helper methods for processing different data.

```java
void handleLdsResponse(DiscoverResponse ldsResponse) {
  checkState((configWatcher != null) != (listenerWatcher != null), ...);  // XOR
  Listener requestedListener = null;
  // unpack Listeners and choose the only one that we are interested in.
  // ...
  if (configWatcher != null) {
    processListenerForConfigUpdate(requestedListener);
  } else {
    processListenerForListenerUpdate(requestedListener);
  }
}

void processListenerForConfigUpdate(Listener listener) {
  // ...
}

void processListenerForListenerUpdate(Listener listener) {
  // ...
}
```

Reply:
Are you sure you want to do this? I don't mind doing that but it is difficult to understand - since the request came from 2 different sources (`ConfigWatcher` vs `ListenerWatcher`) I thought branching right in the beginning might be better/simpler.

Reply:
The logic to locate the requestedListener is different in the 2 cases so it is best to put it in the respective methods (`processListenerForConfigUpdate` vs `processListenerForListenerUpdate`). Done.